storagenode/storagenodedb/storagenodedbtest/multidbsnapshot.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storagenodedbtest
+
+import (
+	"fmt"
+	"sort"
+
+	"storj.io/storj/internal/dbutil/dbschema"
+	"storj.io/storj/internal/dbutil/sqliteutil"
+)
+
+// Snapshots is the global variable that stores all the snapshots for testing
+var Snapshots = MultiDBSnapshots{
+	List: []*MultiDBSnapshot{
+		&v0,
+		&v1,
+		&v2,
+		&v3,
+		&v4,
+		&v5,
+		&v6,
+		&v7,
+		&v8,
+		&v9,
+		&v10,
+		&v11,
+		&v12,
+		&v13,
+		&v14,
+		&v15,
+		&v16,
+		&v17,
+		&v18,
+		&v19,
+		&v20,
+	},
+}
+
+// MultiDBSnapshots provides a convenient list of MultiDBSnapshot
+type MultiDBSnapshots struct {
+	List []*MultiDBSnapshot
+}
+
+// FindVersion finds a snapshot with the specified version.
+func (mdbs *MultiDBSnapshots) FindVersion(version int) (*MultiDBSnapshot, bool) {
+	for _, snap := range mdbs.List {
+		if snap.Version == version {
+			return snap, true
+		}
+	}
+	return nil, false
+}
+
+// Sort sorts the snapshots by version
+func (mdbs *MultiDBSnapshots) Sort() {
+	sort.Slice(mdbs.List, func(i, k int) bool {
+		return mdbs.List[i].Version < mdbs.List[k].Version
+	})
+}
+
+// LoadSnapshots calls LoadSnapshot on each added snapshot
+func (mdbs *MultiDBSnapshots) LoadSnapshots() error {
+	mdbs.Sort()
+	for _, multiDBSnapshot := range mdbs.List {
+		err := multiDBSnapshot.LoadSnapshot()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// MultiDBSnapshot represents an expected state among multiple databases
+type MultiDBSnapshot struct {
+	Version   int
+	Databases Databases
+}
+
+// LoadSnapshot parses the SQL and NewData fields on each database and loads the
+// expected shema and data
+func (mdbs *MultiDBSnapshot) LoadSnapshot() error {
+	for _, database := range mdbs.Databases {
+		snapshot, err := sqliteutil.LoadSnapshotFromSQL(fmt.Sprintf("%s\n%s", database.SQL, database.NewData))
+		if err != nil {
+			return err
+		}
+		database.Schema = snapshot.Schema
+		database.Data = snapshot.Data
+	}
+	return nil
+}
+
+// Databases is a convenience type
+type Databases map[string]*DBSnapshot
+
+// DBSnapshot is a snapshot of a single DB. It separates the SQL and NewData,
+// and after LoadSnapshot is called provides the shema and data.
+type DBSnapshot struct {
+	SQL     string
+	NewData string
+
+	// These are populated by the LoadSnapshot method
+	Schema *dbschema.Schema
```

Comment:
You can embed `*dbschema.Snapshot` here instead of separating them.

I also try to separate the "definition" from the processed thing when performance isn't a concern. This ends up as more code, but it can be clearer, because you don't change the existing entities that could get out of sync.

In this case I was even thinking now, what if we did it when we compare the states, rather than upfront.

Reply:
The main reason it did the loading upfront previously was that it's slightly easier with files.

Reply:
I have separated the definitions. The one that contains the sql I called `MultiDBState`, and the parsed version that has the schema and data I called `MultiDBSnapshot`.

I chose not to embed the `*dbschema.Snapshot`, as it doesn't quite apply: that struct has a version number, stores the script, etc., and I prefer to explicitly name my variables, which I think makes sense in this instance.

I also changed so it does the conversion at checking time, rather than "load" them all at the beginning, as you suggested.

Reply:
I was initially thinking of doing the loading in the comparison, i.e. the test that that https://github.com/storj/storj/pull/3005/files#diff-59b42b6304dea128437983b526d7ee58R123.. but the current approach is fine as well.